push_notifications/wns.py

```diff
@@ -0,0 +1,359 @@
+	data_bytes = bytes(data, "utf-8")
+	headers = {
+		"Content-Type": "application/x-www-form-urlencoded",
+	}
+
+	request = Request(WNS_ACCESS_URL, data=data_bytes, headers=headers)
+	try:
+		response = urlopen(request)
+	except HTTPError as err:
+		if err.code == 400:
+			# One of your settings is probably jacked up.
+			# https://msdn.microsoft.com/en-us/library/windows/apps/xaml/hh868245
+			raise WNSAuthenticationError("Authentication failed, check your WNS settings.")
+		raise err
+	return json.loads(response.read().decode("utf-8"))
+
+
+def _wns_send(uri, data, wns_type="wns/toast"):
+	"""
+	Sends a notification data and authentication to WNS.
+
+	:param uri: str: The device's unique notification URI
+	:param data: dict: The notification data to be sent.
+	:return:
+	"""
+	access_token = _wns_authenticate()["access_token"]
+
+	authorization = "Bearer %(token)s" % {"token": access_token}
+
+	content_type = "text/xml"
+	if wns_type == "wns/raw":
+		content_type = "application/octet-stream"
+
+	headers = {
+		"Content-Type": content_type,  # "text/xml" (toast/badge/tile) | "application/octet-stream" (raw)
+		"Authorization": authorization,
+		"X-WNS-Type": wns_type,  # wns/toast | wns/badge | wns/tile | wns/raw
+	}
+
+	if type(data) is str:
+		data = data.encode('utf-8')
+
+	request = Request(uri, data, headers)
+
+	# A lot of things can happen, let 'em know which one.
+	try:
+		response = urlopen(request)
+	except HTTPError as err:
+		if err.code == 400:
+			raise WNSNotificationResponseError("HTTP 400: One or more headers were specified incorrectly or conflict "
+												"with another header.")
+		elif err.code == 401:
+			raise WNSNotificationResponseError("HTTP 401: The cloud service did not present a valid authentication "
+												"ticket. Access token may be invalid.")
+		elif err.code == 403:
+			raise WNSNotificationResponseError("HTTP 403: The cloud service is not authorized to send a notification to"
+												" this URI even though they are authenticated.")
+		elif err.code == 404:
+			raise WNSNotificationResponseError("HTTP 404: The channel URI is not valid or is not recognized by WNS.")
+		elif err.code == 405:
+			raise WNSNotificationResponseError("HTTP 405: Invalid method (GET, CREATE); only POST (Windows or Windows "
+												"Phone) or DELETE (Windows Phone only) is allowed.")
+		elif err.code == 406:
+			raise WNSNotificationResponseError("HTTP 406: The cloud service exceeded its throttle limit.")
+		elif err.code == 410:
+			raise WNSNotificationResponseError("HTTP 410: The channel expired.")
+		elif err.code == 413:
+			raise WNSNotificationResponseError("HTTP 413: The notification payload exceeds the 5000 byte size limit.")
+		elif err.code == 500:
+			raise WNSNotificationResponseError("HTTP 500: An internal failure caused notification delivery to fail.")
+		elif err.code == 503:
+			raise WNSNotificationResponseError("HTTP 503: The server is currently unavailable.")
```

Comment:
There is really no need for this. You can `raise WNSNotificationResponseError(err)` instead.